Add support for Network Endpoint Group to use serverlessDeployment for API Gateway

kylejohnson514 opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
Addresses https://github.com/hashicorp/terraform-provider-google/issues/10061

Still need to update website docs


<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: Added field `serverless_deployment` to `google_compute_network_endpoint_group` (beta only) for API Gateway resources
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@slevenick, please review this PR or find an appropriate assignee.

kylejohnson514 commented:
I didn't see any `resource_compute_network_endpoint_group.go` or `resource_compute_network_endpoint_group.go.erb` files in the `magic-modules` repo, so I copied over the file from https://github.com/hashicorp/terraform-provider-google/blob/master/google/resource_compute_network_endpoint_group.go and added the conditional imports/usage for code dependent on the beta version of the GCP api's. Was that the right way to do this?

slevenick commented:
> I didn't see any `resource_compute_network_endpoint_group.go` or `resource_compute_network_endpoint_group.go.erb` files in the `magic-modules` repo, so I copied over the file from https://github.com/hashicorp/terraform-provider-google/blob/master/google/resource_compute_network_endpoint_group.go and added the conditional imports/usage for code dependent on the beta version of the GCP api's. Was that the right way to do this?

This resource is generated via the combination of api.yaml and terraform.yaml files here: https://github.com/GoogleCloudPlatform/magic-modules/tree/master/mmv1/products/compute

The right way to add this would be to add the serverlessDeploy field to the api.yaml

slevenick reviewed (changes requested):
This resource is generated, so additional fields should be added to the api.yaml for the resource

kylejohnson514 commented:
Did those updates, is that all? Will `flattenComputeNetworkEndpointGroupServerlessDeployment` and `expandComputeNetworkEndpointGroupServerlessDeployment` be autogenerated to use the https://pkg.go.dev/google.golang.org/api/compute/v0.beta#NetworkEndpointGroupServerlessDeployment object?

slevenick commented:
You will need to remove the resource file that you copied from the downstream provider. That file is automatically generated during the generate-diffs process, and the flatten & expand methods will be auto-generated as well.

You can tag new fields with `min_version: beta` to add them only to the beta provider

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 82 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 232 insertions(+), 9 deletions(-))

ivoilic commented:
@slevenick @kylejohnson514 What else needs to be done to get this PR clear to merge? Happy to help, if I can.

slevenick commented on mmv1/third_party/terraform/tests/resource_compute_network_endpoint_group_test.go:
```diff
@@ -31,6 +31,33 @@ func TestAccComputeNetworkEndpointGroup_networkEndpointGroup(t *testing.T) {
 	})
 }
 
+<% unless version == 'ga' -%>
```
This file needs to be renamed to end with `.go.erb` now that it has embedded ruby in it.

slevenick reviewed (changes requested):
Failing to build due to ruby tags in a `.go` file. Changing the file name should fix this

slevenick commented:
/gcbrun

kylejohnson514 commented:
Hi, any updates on this? Would like to get this finished up and merged

ivoilic commented:
Hey @rileykarson would you mind taking a look at this? There are a lot of issues requesting the feature in this PR over on the terraform repo. It would be great to get some feedback on @kylejohnson514 's work so it can get merged in.

slevenick commented:
> Hi, any updates on this? Would like to get this finished up and merged

Yeah, sorry. Looks like the PR is failing to build due to a missing autogen notice in a new file:

`/mm-terraform-ga-head/mmv1/provider/file_template.rb:69:in `generate': google/resource_compute_network_endpoint_group_test.go missing autogen (RuntimeError)`

The file will need an autogen exception added at the top as it is now a template. This looks like: `<% autogen_exception -%>`

Are you able to build the Terraform provider yourself from this branch?

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 45 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 240 insertions(+), 9 deletions(-))

kylejohnson514 commented:
> Are you able to build the Terraform provider yourself from this branch?

Yep, I figured out building the provider locally, and your fix was correct.

I cleaned up the tests because I meant to be using a `RegionNetworkEndpointGroup` object but originally was trying to test via `NetworkEndpointGroup`. I also updated the resources in my unit test to use Cloud Run, and based it off of https://github.com/hashicorp/terraform-provider-google-beta/blob/6f217d8a45a4eb639200fbf440848ad76c415d4f/google-beta/resource_compute_region_network_endpoint_group_generated_test.go#L115 .

But, running into this error now...
```
~/go/src/github.com/hashicorp/terraform-provider-google-beta master* ❯ make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment'      04:36:12 PM
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment
=== PAUSE TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment
=== CONT  TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment
    provider_test.go:286: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error creating RegionNetworkEndpointGroup: googleapi: Error 400: Invalid value for field 'resource.id': '0'. Serverless Deployment platform 'run.googleapis.com' is not supported yet., invalid
        
          with google_compute_region_network_endpoint_group.cloudrun_serverless_neg,
          on terraform_plugin_test.tf line 2, in resource "google_compute_region_network_endpoint_group" "cloudrun_serverless_neg":
           2: resource "google_compute_region_network_endpoint_group" "cloudrun_serverless_neg" {
        
--- FAIL: TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment (23.10s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta 24.248s
FAIL
make: *** [testacc] Error 1
```

Any idea about this? According to https://cloud.google.com/compute/docs/reference/rest/beta/regionNetworkEndpointGroups we should be fine using with `platform = "run.googleapis.com"`

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 45 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 253 insertions(+), 9 deletions(-))

slevenick commented:
Looks like run.googleapis.com is invalid despite what the docs say. I checked around internally and only the only accepted value for the platform field is apigateway.googleapis.com, and only in the beta API. The supported way to bring up a Cloud Run NEG is using the cloudRun field on the network endpoint group.

kylejohnson514 commented:
Now spinning up a new API Gateway resource and trying to connect the serverless NEG to it. I can see the resources get spun up correctly in GCP, but it fails when it gets to the actual NEG. Error doesn't really make sense to me:

```
make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment
=== PAUSE TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment
=== CONT  TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment
    provider_test.go:286: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error creating RegionNetworkEndpointGroup: googleapi: Error 400: Invalid value for field 'resource.name': 'test-serverless-neg'. URL mask must contain at least one valid label. Valid label(s): [gateway]., invalid
        
          with google_compute_region_network_endpoint_group.test_neg,
          on terraform_plugin_test.tf line 27, in resource "google_compute_region_network_endpoint_group" "test_neg":
          27: resource "google_compute_region_network_endpoint_group" "test_neg" {
        
--- FAIL: TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment (477.91s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta 478.947s
FAIL
make: *** [testacc] Error 1
```

The only part of the docs that mention a label requirement (i.e. tag/value) is when you're spinning up Cloud Run resources, but Cloud Run isn't actually supported in the beta yet. I'm also trying to spin up API Gateway anyways, so the Cloud Run requirement shouldn't matter.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 44 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 258 insertions(+), 9 deletions(-))

slevenick commented:
Got this answer for you:

The URL mask is invalid. It should look something like "https://<gateway>.myloadbalancer.com/" or "https://myloadbalancer.com/<gateway>". The idea is to give it a string by which the LB can extract the identity of the gateway being requested from the URL.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 44 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 258 insertions(+), 9 deletions(-))

slevenick commented:
/gcbrun

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 44 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 257 insertions(+), 9 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 44 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 257 insertions(+), 9 deletions(-))

modular-magician commented:
I have triggered VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccApigeeEnvironmentIamBindingGenerated|TestAccApigeeEnvironmentIamPolicyGenerated|TestAccCloudbuildWorkerPool_basic|TestAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment|TestAccContainerNodePool_withInvalidUpgradeSettings You can view the result here: https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=246759

slevenick commented:
Looks like we have a couple failures due to linting:

```
google/resource_compute_region_network_endpoint_group_test.go:4:2: imported and not used: "testing"
google/resource_compute_region_network_endpoint_group_test.go:6:2: imported and not used: "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
```

Other than that the tests pass, so once those are fixed I'll do a final pass and it should be good to go

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 38 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 257 insertions(+), 9 deletions(-))

slevenick commented:
/gcbrun

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 38 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 257 insertions(+), 9 deletions(-))

modular-magician commented:
I have triggered VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccDatasourceGoogleServiceNetworkingPeeredDnsDomain_basic|TestAccApigeeEnvironmentIamBindingGenerated|TestAccApigeeEnvironmentIamMemberGenerated|TestAccApigeeEnvironmentIamPolicyGenerated|TestAccCloudbuildWorkerPool_basic|TestAccServiceNetworkingPeeredDNSDomain_basic You can view the result here: https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=248106

slevenick commented on mmv1/third_party/terraform/tests/resource_compute_region_network_endpoint_group_test.go.erb:
```diff
@@ -0,0 +1,78 @@
+func testAccComputeRegionNetworkEndpointGroup_negWithServerlessDeployment(context map[string]interface{}) string {
+	return Nprintf(`
+resource "google_api_gateway_api" "api_gw" {
+	api_id = "api-gw"
```
Can we change all the user set ID and names in this test config to the format:

`tf-test-%{random_suffix}`? It allows our CI that runs tests to delete old resources that are left around, and not run into issues with conflicting names if this test is run twice at the same time

kylejohnson514 replied:
Fixed

slevenick reviewed (changes requested):
Just a small change to test files, otherwise looks fine!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 38 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 257 insertions(+), 9 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 38 insertions(+), 6 deletions(-))

kylejohnson514 commented:
Is there anything else to change @slevenick ?

slevenick commented:
/gcbrun

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 38 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5434-old..auto-pr-5434) ( 3 files changed, 257 insertions(+), 9 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5434-old..auto-pr-5434) ( 2 files changed, 3 insertions(+), 3 deletions(-))

slevenick reviewed (approved):
Thanks!

jose-lpa commented on mmv1/third_party/terraform/tests/resource_compute_region_network_endpoint_group_test.go.erb:
```diff
@@ -0,0 +1,78 @@
+	region                = "us-central1"
+	serverless_deployment {
+		platform = "apigateway.googleapis.com"
+		url_mask = format("<gateway>%s/hello", trimprefix(google_api_gateway_gateway.api_gw.default_hostname, "tf-test-%{random_suffix}"))
```
I couldn't find this thing documented anywhere, and it was quite difficult for me to extract from the docs, both Google GCP and Terraform, what value needs to be put in here for the serverless API Gateway deployment.

hanzala1234 replied:
faced the same thing. I think it should be documented somewhere
